apis/v1alpha2/gateway_types.go

```diff
@@ -0,0 +1,835 @@
+	//   or the [:authority pseudo-header](https://tools.ietf.org/html/rfc7540#section-8.1.2.3)
+	// * For the "HTTPS" protocol, the Hostname match MUST be
+	//   applied at both the TLS and HTTP protocol layers.
+	//
+	// Support: Core
+	Protocol ProtocolType `json:"protocol"`
+
+	// TLS is the TLS configuration for the Listener. This field
+	// is required if the Protocol field is "HTTPS" or "TLS" and
+	// ignored otherwise.
+	//
+	// The association of SNIs to Certificate defined in GatewayTLSConfig is
+	// defined based on the Hostname field for this listener.
+	//
+	// The GatewayClass MUST use the longest matching SNI out of all
+	// available certificates for any TLS handshake.
+	//
+	// Support: Core
+	//
+	// +optional
+	TLS *GatewayTLSConfig `json:"tls,omitempty"`
+
+	// Routes specifies which Routes may be attached to this Listener.
+	//
+	// Although a client request may technically match multiple route rules,
+	// only one rule may ultimately receive the request. Matching precedence
+	// MUST be determined in order of the following criteria:
+	//
+	// * The most specific match. For example, the most specific HTTPRoute match
+	//   is determined by the longest matching combination of hostname and path.
+	// * The oldest Route based on creation timestamp. For example, a Route with
+	//   a creation timestamp of "2020-09-08 01:02:03" is given precedence over
+	//   a Route with a creation timestamp of "2020-09-08 01:02:04".
+	// * If everything else is equivalent, the Route appearing first in
+	//   alphabetical order (namespace/name) should be given precedence. For
+	//   example, foo/bar is given precedence over foo/baz.
+	//
+	// All valid portions of a Route selected by this field should be supported.
+	// Invalid portions of a Route can be ignored (sometimes that will mean the
+	// full Route). If a portion of a Route transitions from valid to invalid,
+	// support for that portion of the Route should be dropped to ensure
+	// consistency. For example, even if a filter specified by a Route is
+	// invalid, the rest of the Route should still be supported.
+	//
+	// Support: Core
+	// +kubebuilder:default={namespaces:{from: Same}}
+	// +optional
+	Routes *ListenerRoutes `json:"routes,omitempty"`
+}
+
+// ProtocolType defines the application protocol accepted by a Listener.
+// Implementations are not required to accept all the defined protocols.
+// If an implementation does not support a specified protocol, it
+// should raise a "Detached" condition for the affected Listener with
+// a reason of "UnsupportedProtocol".
+//
+// Core ProtocolType values are listed in the table below.
+//
+// Implementations can define their own protocols if a core ProtocolType does not
+// exist. Such definitions must use prefixed name, such as
+// `mycompany.com/my-custom-protocol`. Un-prefixed names are reserved for core
+// protocols. Any protocol defined by implementations will fall under custom
+// conformance.
+type ProtocolType string
+
+const (
```

Comment:
This is probably why it is so confusing. To me, that is. We are mixing L7 Protos with protocol semantics such as HTTP vs HTTPs vs `TLS` which is a semantic for both TCP and HTTP.

Reply:
Discussed at community meeting, @bowei will work on clarifying docs here. Maybe ListenerType makes more sense than ProtocolType here?

Reply:
`ProtocolType` here isn't strictly layered (if you go down that path, you end up with a lot of complexity and semantic wrangling). Think of it as a name for protocol at the layer that you want to deal with it on.
For example, "HTTPS", instead of "IPV{4,6} + {TCP,UDP} + {HTTP/1.1,HTTP/2,HTTP/3}".

But because `ProtocolType` is not strict, implementations are also free to add their own types, with few constraints about what those types ought to mean.